From b9684305bab53c6c9705ccfaf669c118d9bb501c Mon Sep 17 00:00:00 2001
From: Lukas Czerner <lczerner@redhat.com>
Date: Wed, 29 Oct 2014 16:34:24 -0400
Subject: [fs] jbd2: fix journal checksum feature flag handling

Message-id: <1414600485-1929-35-git-send-email-lczerner@redhat.com>
Patchwork-id: 98557
O-Subject: [RHEL 7.1 PATCH v2 34/55] jbd2: fix journal checksum feature flag handling
Bugzilla: 1156096
RH-Acked-by: Eric Sandeen <sandeen@redhat.com>

From: "Darrick J. Wong" <darrick.wong@oracle.com>

BZ 1156096
https://bugzilla.redhat.com/show_bug.cgi?id=1156096

BREW 8169722
https://brewweb.devel.redhat.com/taskinfo?taskID=8169722

Upstream commit feb8c6d3dd0f2cc0e1c3376d099cf298c5f2c2c8

 Clear all three journal checksum feature flags before turning on
 whichever journal checksum options we want.  Rearrange the error
 checking so that newer flags get complained about first.

 Reported-by: TR Reardon <thomas_reardon@hotmail.com>
 Signed-off-by: Darrick J. Wong <darrick.wong@oracle.com>
 Signed-off-by: Theodore Ts'o <tytso@mit.edu>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/ext4/super.c b/fs/ext4/super.c
index 48ac819..161d613 100644
--- a/fs/ext4/super.c
+++ b/fs/ext4/super.c
@@ -3236,6 +3236,10 @@ static int set_journal_csum_feature_set(struct super_block *sb)
   incompat = 0;
  }
 
+ jbd2_journal_clear_features(sbi->s_journal,
+   JBD2_FEATURE_COMPAT_CHECKSUM, 0,
+   JBD2_FEATURE_INCOMPAT_CSUM_V3 |
+   JBD2_FEATURE_INCOMPAT_CSUM_V2);
  if (test_opt(sb, JOURNAL_ASYNC_COMMIT)) {
   ret = jbd2_journal_set_features(sbi->s_journal,
     compat, 0,
@@ -3248,11 +3252,8 @@ static int set_journal_csum_feature_set(struct super_block *sb)
   jbd2_journal_clear_features(sbi->s_journal, 0, 0,
     JBD2_FEATURE_INCOMPAT_ASYNC_COMMIT);
  } else {
-  jbd2_journal_clear_features(sbi->s_journal,
-    JBD2_FEATURE_COMPAT_CHECKSUM, 0,
-    JBD2_FEATURE_INCOMPAT_ASYNC_COMMIT |
-    JBD2_FEATURE_INCOMPAT_CSUM_V3 |
-    JBD2_FEATURE_INCOMPAT_CSUM_V2);
+  jbd2_journal_clear_features(sbi->s_journal, 0, 0,
+    JBD2_FEATURE_INCOMPAT_ASYNC_COMMIT);
  }
 
  return ret;
diff --git a/fs/jbd2/journal.c b/fs/jbd2/journal.c
index 19d74d8..7e70cd5 100644
--- a/fs/jbd2/journal.c
+++ b/fs/jbd2/journal.c
@@ -1522,14 +1522,6 @@ static int journal_get_superblock(journal_t *journal)
   goto out;
  }
 
- if (jbd2_journal_has_csum_v2or3(journal) &&
-     JBD2_HAS_COMPAT_FEATURE(journal, JBD2_FEATURE_COMPAT_CHECKSUM)) {
-  /* Can't have checksum v1 and v2 on at the same time! */
-  printk(KERN_ERR "JBD2: Can't enable checksumming v1 and v2 "
-         "at the same time!\n");
-  goto out;
- }
-
  if (JBD2_HAS_INCOMPAT_FEATURE(journal, JBD2_FEATURE_INCOMPAT_CSUM_V2) &&
      JBD2_HAS_INCOMPAT_FEATURE(journal, JBD2_FEATURE_INCOMPAT_CSUM_V3)) {
   /* Can't have checksum v2 and v3 at the same time! */
@@ -1538,6 +1530,14 @@ static int journal_get_superblock(journal_t *journal)
   goto out;
  }
 
+ if (jbd2_journal_has_csum_v2or3(journal) &&
+     JBD2_HAS_COMPAT_FEATURE(journal, JBD2_FEATURE_COMPAT_CHECKSUM)) {
+  /* Can't have checksum v1 and v2 on at the same time! */
+  printk(KERN_ERR "JBD2: Can't enable checksumming v1 and v2/3 "
+         "at the same time!\n");
+  goto out;
+ }
+
  if (!jbd2_verify_csum_type(journal, sb)) {
   printk(KERN_ERR "JBD2: Unknown checksum type\n");
   goto out;
-- 
1.7.1